sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -266,6 +266,13 @@ trait CheckAnalysis extends PredicateHelper {
                 s"but one table has '${firstError.output.length}' columns and another table has " +
                 s"'${s.children.head.output.length}' columns")
 
+          case Union(c) if c.exists(_.output.map(_.dataType) != c.head.output.map(_.dataType)) =>
```

Comment:
This is a bit messy. Can we just move this into the match body and combine it with firstError?